src/aoc_cj/aoc2024/day16.py

```diff
@@ -0,0 +1,101 @@
+import dataclasses
+import enum
+import itertools
+import math
+from collections import deque
+from collections.abc import Callable, Generator
+from typing import Self
+
+import more_itertools as mi
+
+from aoc_cj import util
+
+
+class Facing(enum.Enum):
+    NORTH = -1j
+    SOUTH = 1j
+    EAST = 1
+    WEST = -1
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True, slots=True)
+class _State:
+    pos: complex
+    facing: Facing
+
+    def next(self, valid_pos: Callable[[complex], bool]) -> Generator[tuple[Self, int], None, None]:
+        cls = type(self)
+        for new_pos in util.adj_4(self.pos):
+            if valid_pos(new_pos):
+                new_facing = Facing(new_pos - self.pos)
+                turned = new_facing != self.facing
+                cost_increase = (1000 if turned else 0) + 1
+                yield cls(pos=new_pos, facing=new_facing), cost_increase
+
+
+def part_1(txt: str) -> int:
+    start, end, walls = parse(txt)
+
+    initial = _State(pos=start, facing=Facing.EAST)
+    to_explore = deque(((0, initial),))
+    valid_pos = lambda p: p not in walls
+    cheapest: dict[_State, int] = {}
+    while to_explore:
+        cost, state = to_explore.popleft()
+        if cost < cheapest.get(state, math.inf):
+            cheapest[state] = cost
+            for next_state, cost_increase in state.next(valid_pos):
+                to_explore.append((cost + cost_increase, next_state))
+
+    return min(cost for s, cost in cheapest.items() if s.pos == end)
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True, slots=True)
+class _CheapTracker:
+    cost: int
+    seen: set[complex]
+
+
+def part_2(txt: str) -> int:
+    start, end, walls = parse(txt)
+
+    # TODO: there's a lot of duplicated stuff. can we refactor?
```

Comment:
only so many hours in the day. If I needed a more general solution, I'd find (or write) a bfs/dfs library, but this is good enough to solve the AoC problem.

```suggestion
```